pkg/kv/kvserver/store.go

```diff
@@ -2192,70 +2191,108 @@ func (s *Store) startLeaseRenewer(ctx context.Context) {
+			// closed timestamp lag for each range. The closedts-rangefeed-updater
+			// refresh loop operates independently of incoming closed timestamp
+			// updates, so this does not favor some replicas over others.
+			sort.Sort(&rangeIDs)
+			return rangeIDs
+		}
 
-			getInterval := func() time.Duration {
-				refresh := RangeFeedRefreshInterval.Get(&st.SV)
-				if refresh != 0 {
-					return refresh
-				}
-				return closedts.SideTransportCloseInterval.Get(&st.SV)
+		timer := timeutil.NewTimer()
+		defer timer.Stop()
+		errInterrupted := errors.New("waiting interrupted")
+		wait := func(ctx context.Context, until time.Time, interrupt <-chan struct{}) error {
+			now := timeutil.Now()
+			if !now.Before(until) {
+				return nil
+			}
+			timer.Reset(until.Sub(now))
+			select {
+			case <-timer.C:
+				timer.Read = true
+				return nil
+			case <-interrupt:
+				return errInterrupted
+			case <-ctx.Done():
+				return ctx.Err()
 			}
+		}
 
-			for {
-				interval := getInterval()
-				if interval > 0 {
-					timer.Reset(interval)
-				} else {
-					// Disable the side-transport.
-					timer.Stop()
-					timer = timeutil.NewTimer()
+		// Repeatedly handle the closed timestamp update for each range ID.
+		//
+		// The `refresh` interval is the target interval between delivering closed
+		// timestamp updates for a single range. This is the target time in which a
+		// single pass of the loop below should complete. The `smear` interval is
+		// the target interval between wake-ups within a single run.
+		//
+		// We smear the work across the refresh interval to avoid waking up all the
+		// replicas' rangefeed processor and registration goroutines in a short
+		// burst. Doing so without pacing can have a severe impact on goroutine
+		// scheduling latency which impacts foreground latency, as tens of thousands
+		// of goroutines can be marked as runnable in a short period of time.
+		//
+		// Pacing works as follows:
+		// - We spread closed timestamp notifications to replicas across the refresh
+		//   interval (default 200ms).
+		// - We smear across replicas at the granularity of smearInterval (e.g. 1ms)
+		//   to avoid putting this goroutine to sleep for excessively short periods
+		//   of time.
+		// - We prioritize staying within the target refresh interval, so we adjust
+		//   the rate of processing in accordance with the time remaining until the
+		//   refresh interval ends.
+		conf := newRangeFeedUpdaterConf(s.cfg.Settings)
+		for {
+			// Configuration may have changed between runs, load it unconditionally.
+			refresh, smear, err := conf.wait(ctx)
+			if err != nil {
+				return // context canceled
+			}
+			// Aim to complete this run in exactly refresh interval.
+			now := timeutil.Now()
+			target := now.Add(refresh)
+			var waitErr error
+			for work, startAt := updateRangeIDs(), now; len(work) != 0; {
+				if waitErr = wait(ctx, startAt, conf.changed); waitErr != nil {
```

Comment:
Is it ok that we ignore the tail of pending replicas if conf change arrives and restart notifications? Should be ok in practice, but maybe point that in comment that when configuration changes we abandon current ts update loop and start from scratch (instead of the comment at the end of main for-loop that currently states obvious).

Reply:
Yes, I think it's ok. If we set a large interval (e.g. 1h), and then lower it to 1s, we don't want to wait the whole hour (until the current run completes) before the config change takes effect.

We could honour the new parameters without restarting the run though. The `rangeFeedUpdaterPace` makes the algorithm adaptive, we could change `deadline` and `smear` interval on the fly. Added a TODO.